README.md
# Enriched UDP Socket and Client 
[![npm](https://img.shields.io/npm/v/e-socket-udp)](https://www.npmjs.com/package/e-socket-udp)
[![tests](https://github.com/JerryCauser/e-socket-udp/actions/workflows/tests.yml/badge.svg)](https://github.com/JerryCauser/e-socket-udp/actions/workflows/tests.yml)
[![CodeQL](https://github.com/JerryCauser/e-socket-udp/actions/workflows/codeql.yml/badge.svg)](https://github.com/JerryCauser/e-socket-udp/actions/workflows/codeql.yml)
[![CodeFactor Grade](https://img.shields.io/codefactor/grade/github/JerryCauser/e-socket-udp/master)](https://www.codefactor.io/repository/github/jerrycauser/e-socket-udp)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)
[![node-current](https://img.shields.io/node/v/e-socket-udp)](https://nodejs.org)
[![GitHub](https://img.shields.io/github/license/JerryCauser/e-socket-udp)](https://github.com/JerryCauser/e-socket-udp/blob/master/LICENSE)

Enriched UDP Socket and Client with possibility to encrypt data and send big messages.

- Fast — small overhead above UDP to send messages
- Secure — built-in encryption for sensitive logs
- Universal – built-in fragmentation to send big messages
- Simple — used well-known Node streams to manipulate and move data 
- ESM and CJS

## Install

```bash
npm i --save e-socket-udp
```

## Fast Start

```javascript
//client.js
import { UDPClient } from 'e-socket-udp'

const secret = '11223344556677889900aabbccddeeff11223344556677889900aabbccddeeff'
const client = new UDPClient({ encryption: secret })

client.write(Buffer.from('Hello, World!', 'utf8'))
```

```javascript
//server.js
import { UDPSocket } from 'e-socket-udp'

const secret = '11223344556677889900aabbccddeeff11223344556677889900aabbccddeeff'
const socket = new UDPSocket({ decryption: secret })

for await (const message of socket) {
  console.log(message)
}
```

After just start the server  `node server.js` and client `node app.js`. That's all and everything works.

## Documentation

### class `UDPClient`
Extends [Basic UDPClient][basic-udp-client].

#### Arguments:
- `options` `<object>` – optional
  - `type` `<'udp4' | 'udp6'>` – optional. _Inherited_. **Default** `'udp4'`
  - `port` `<string | number>` – optional. _Inherited_. **Default** `44302`
  - `address` `<string>` – optional. _Inherited_. **Default** `'127.0.0.1'` or `'::1'`
  - `encryption` `<string> | <(payload: Buffer) => Buffer>` – optional. **Default** `undefined`
    - if passed `string` - will be applied `aes-256-ctr` encryption with passed string as a secret, so it should be `64char` long;
    - if passed `function` - this function will be used to encrypt every message;
    - if passed `undefined` - will not use any kind of encryption.
  - `fragmentation` `<boolean>` – optional. Automatically split each message into chunks. Useful with any size of messages (especially big ones). **Default** `true`  
  - `packetSize` `<number>` – optional. Used if `fragmentation = true`. Number of bytes in each packet (chunk). **Default** `1280`

#### Methods:
- `write (buffer)`: `<boolean>` — _Inherited_.

#### Fields:
- `origin`: [`<dgram.Socket>`][node-dgram-socket] — _Inherited_.
- `port`: `<number>` — _Inherited_.
- `address`: `<string>` — _Inherited_.
- `family`: `<string>` — _Inherited_.

#### Events:
- `'ready'`: `<void>` – _Inherited_. Emitted when the client "establishes" udp connection.

#### Usage
##### Simple example with encryption
```javascript
import { UDPClient } from 'e-socket-udp'

const client = new UDPClient({
  port: 44302,
  // encryption: (buf) => buf.map(byte => byte ^ 83)
  encryption: '11223344556677889900aabbccddeeff11223344556677889900aabbccddeeff'
})

client.write(Buffer.from('Hello, world!', 'utf8'))
```

---

### class `UDPSocket`
Extends [Basic UDPSocket][basic-udp-socket]

It is a UDP socket in `readable stream` form with encoding and possibility to handle big messages.

#### Arguments:
- `options` `<object>` – **required**
  - `type` `<'udp4' | 'udp6'>` – optional. _Inherited_. **Default** `'udp4'`
  - `port` `<string | number>` – optional. _Inherited_. **Default** `44302`
  - `host` `<string>` – optional. _Inherited_. **Default** `'127.0.0.1'` or `'::1'`
  - `decryption` `<string> | <(payload: Buffer) => Buffer>` – optional. **Default** `undefined`
    - if passed `string` - will be applied `aes-256-ctr` decryption with passed string as a secret, so it should be `64char` long;
    - if passed `function` - will be used that function to decrypt every message;
    - if passed `undefined` - will not use any kind of decryption.
  - `fragmentation` `<boolean>` – optional. Combine chunks into message. Useful with any size of messages (especially big ones). **Default** `true`
  - `gcIntervalTime` `<number>` — optional. How often instance will check internal buffer to delete expired messages (in ms). **Default** `5000` 
  - `gcExpirationTime` `<number>`— optional. How long chunks can await all missing chunks in internal buffer (in ms). **Default** `10000`

#### Fields:
- `origin`: [`<dgram.Socket>`][node-dgram-socket] — _Inherited_.
- `port`: `<number>` — _Inherited_.
- `address`: `<string>` — _Inherited_.
- `family`: `<string>` — _Inherited_.
- `allowPush`: `<boolean>` — _Inherited_.

#### Events:
All `Readable` events of course and:

##### Event: `'ready'` — _Inherited_.
Emitted when socket started and ready to receive data.

##### Event: `'warning'`
Emitted when warning occurs.
 - `message` `<object | Error>`
   - `type` `<Symbol>`
   - `id` `<string>` – optional
   - `date` `<Date>` – optional

A type might be:
   - `Symbol<'missing_message'>` – when some messages didn't receive all chunks and got expired.
   - `Symbol<'decryption_message'>` – when some messages failed to be compiled from chunks.

#### Usage
##### Example how to use pure socket as async generator
```javascript
import { UDPSocket } from 'e-socket-udp'

const socket = new UDPsocket({
  port: 44302,
  // decryption: (buf) => buf.map(byte => byte ^ 83) // not safe at all, but better than nothing
  decryption: '11223344556677889900aabbccddeeff11223344556677889900aabbccddeeff'
})

for await (const message of socket) {
  /*handle messages*/
}
```

##### Example how to use socket and fs write stream
```javascript
import fs from 'node:fs'
import { UDPSocket } from 'e-socket-udp'

const secret = '11223344556677889900aabbccddeeff11223344556677889900aabbccddeeff'

const writer = fs.createWriteStream('/some/path')
const socket = new UDPSocket({ port: 44302, decryption: secret })

socket.pipe(writer)
```

---


### Additional Exposed variables and functions
#### constant `DEFAULT_PORT`
- `<number>` : `44302`

#### constant `WARNING_MISSING_MESSAGE`
- `<Symbol>` : `Symbol<'missing_message'>`

#### constant `WARNING_DECRYPTION_FAIL`
- `<Symbol>` : `Symbol<'decryption_fail'>`
---

There are `_identifier` and `_constants` exposed also, but they are used for internal needs. They could be removed in next releases, so it is not recommended to use it in your project.  

---

License ([MIT](LICENSE))

[basic-udp-client]: https://github.com/JerryCauser/socket-udp#class-udpclient
[basic-udp-socket]: https://github.com/JerryCauser/socket-udp#class-udpsocket
[node-readable]: https://nodejs.org/api/stream.html#class-streamreadable
[node-writable]: https://nodejs.org/api/stream.html#class-streamwritable
[node-dgram-socket]: https://nodejs.org/api/dgram.html#class-dgramsocket
[client]: #class-udpclient
[socket]: #class-udpsocket
[constants]: src/constants.js
[socket-event-warning]: #event-warning


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.3](https://github.com/JerryCauser/e-socket-udp/compare/v0.3.2...v0.3.3) (2023-01-09)


### Bug Fixes

* sent interrupted messages after drain ([3439f8d](https://github.com/JerryCauser/e-socket-udp/commit/3439f8dc23e3681ee6628a8f362bedcc0e885422))

### [0.3.2](https://github.com/JerryCauser/e-socket-udp/compare/v0.3.1...v0.3.2) (2023-01-03)

### [0.3.1](https://github.com/JerryCauser/e-socket-udp/compare/v0.3.0...v0.3.1) (2023-01-02)

## [0.3.0](https://github.com/JerryCauser/e-socket-udp/compare/v0.2.0...v0.3.0) (2023-01-01)

## [0.2.0](https://github.com/JerryCauser/e-socket-udp/compare/v0.1.1...v0.2.0) (2023-01-01)


### Bug Fixes

* remove unused variable ([30c0a19](https://github.com/JerryCauser/e-socket-udp/commit/30c0a1926755c337f13d5bfcb14cc445a26cd147))

### 0.1.1 (2022-12-29)


### Features

* initial code add ([3ea42aa](https://github.com/JerryCauser/e-socket-udp/commit/3ea42aa54546f260bdde6eca1718bccd4bf063da))


tests/_main.js
import identifierTests from './identifier.test.js'
import constantsTest from './constants.test.js'
import clientTest from './client.test.js'
import socketTest from './socket.test.js'
import synergyTest from './synergy.test.js'

/**
 * sequence:
 *  [x] identifier
 *  [x] defaults
 *  [x] client
 *  [x] socket
 */

export default async function _main (
  type,
  { UDPClient, UDPSocket, _constants, _identifier }
) {
  console.log(`${type} Tests Started\n`)
  let errorsCount = 0

  errorsCount += await identifierTests(_identifier)
  errorsCount += await constantsTest(_constants)
  errorsCount += await clientTest(UDPClient, _identifier)
  errorsCount += await socketTest(UDPSocket, _identifier, _constants)
  errorsCount += await synergyTest(UDPSocket, UDPClient)

  if (errorsCount === 0) console.log('All tests passed')
  else {
    throw new Error(`Not all tests are passed. FAILED tests: ${errorsCount}`)
  }
}

/**
 * @returns {{count: number, try: (((fn: function) => Promise<void>))}}
 */
export const tryCountErrorHook = () => {
  const tryCountError = async (fn) => {
    try {
      const call = fn()

      if (call?.then) {
        await call
      }
    } catch (e) {
      ++obj.count
      console.error(e)
    }
  }

  const obj = {
    count: 0,
    try: tryCountError
  }

  return obj
}

/**
 * @param {function} fn
 * @param {object} obj
 */
export const assertTry = (fn, obj) => {
  try {
    fn()
  } catch (e) {
    obj.fails.push(e)
  }
}

/**
 * @param {{fails:Error[]|object[]}} results
 * @param {string} caseAlias
 * @param {any?} ctx
 */
export const checkResults = (results, caseAlias, ctx) => {
  if (results.fails.length > 0) {
    const error = new Error(
      `${caseAlias} failed some tests: Number: ${results.fails.length}. Check description`
    )

    if (ctx) error.data = ctx

    Object.assign(error, results)

    throw error
  } else {
    console.log(`${caseAlias} passed`)
  }
}


tests/client.test.js
import crypto from 'node:crypto'
import assert from 'node:assert'
import dgram from 'node:dgram'
import { once } from 'node:events'
import { tryCountErrorHook } from './_main.js'

/**
 * [x] here we need to create pure datagram server and handle message from client
 * [x] it should be sent and received almost at same time
 * [x] date info should be correct
 * [x] it should contain the correct data for single message
 * [x] separated messages should be correctly joined
 * [x] index and total should be correct
 */

const delay = (ms) => new Promise((resolve) => setTimeout(resolve, ms))

/**
 * @param {number} port
 * @returns {Promise<Socket & {messages:Buffer[], stop: (() => Promise<void>)}>}
 */
const createUDPSocket = async (port) => {
  const socket = Object.create(
    dgram.createSocket({ type: 'udp4', reuseAddr: true })
  )
  socket.bind(port, '127.0.0.1')

  const error = await Promise.race([
    once(socket, 'listening'),
    once(socket, 'error')
  ])

  socket.messages = []

  socket.on('message', (buffer) => socket.messages.push(buffer))

  if (error instanceof Error) throw Error

  socket.stop = async () => {
    socket.removeAllListeners()
    socket.close()
    await once(socket, 'close')
  }

  return socket
}

async function clientTest (UDPClient, identifier) {
  const { ID_SIZE, parseId } = identifier
  const alias = '  client.js:'

  const PAYLOAD_SIZE = 300
  const DEFAULT_PORT = 45002

  async function testClientSmall (port = DEFAULT_PORT, fragmentation) {
    const caseAlias = `${alias} client sending small message fragmentation=${
      fragmentation ? 'true' : 'false'
    } ->`
    const socket = await createUDPSocket(port)
    const client = new UDPClient({
      port,
      fragmentation,
      packetSize: PAYLOAD_SIZE
    })
    const payload = crypto.randomBytes(PAYLOAD_SIZE - ID_SIZE)

    await once(client, 'ready')

    const dateBeforeSent = Date.now()
    client.write(payload)

    await delay(5)

    assert.strictEqual(
      socket.messages.length,
      1,
      `${caseAlias} 1 message should be received by socket`
    )

    const [messageDate, , total, index] = fragmentation
      ? parseId(socket.messages[0].subarray(0, ID_SIZE))
      : []

    fragmentation &&
      assert.ok(
        messageDate >= dateBeforeSent && messageDate <= Date.now(),
        `${caseAlias} Message date invalid ${new Date().toISOString()} but should be ${new Date(
          dateBeforeSent
        ).toISOString()}±5ms`
      )

    fragmentation &&
      assert.strictEqual(
        total,
        0,
        `${caseAlias} Message total isn't same as expected`
      )
    fragmentation &&
      assert.strictEqual(
        index,
        0,
        `${caseAlias} Message index isn't same as expected`
      )

    assert.deepStrictEqual(
      fragmentation ? socket.messages[0].subarray(ID_SIZE) : socket.messages[0],
      payload,
      `${caseAlias} received message should be the same as sent one`
    )

    await socket.stop()

    console.log(`${caseAlias} passed`)
  }

  async function testClientLarge (port = DEFAULT_PORT) {
    const caseAlias = `${alias} client sending large message ->`
    const socket = await createUDPSocket(port)
    const client = new UDPClient({
      port,
      fragmentation: true,
      packetSize: PAYLOAD_SIZE
    })
    const payload = crypto.randomBytes((PAYLOAD_SIZE - ID_SIZE) * 2)

    await once(client, 'ready')

    client.write(payload)

    await delay(5)

    assert.strictEqual(
      socket.messages.length,
      2,
      `${caseAlias} 2 messages should be received by socket`
    )

    // eslint-disable-next-line no-unused-vars
    const [msgOneDate, id1, msgOneTotal, msgOneIndex] = parseId(
      socket.messages[0].subarray(0, ID_SIZE)
    )

    assert.strictEqual(
      msgOneTotal,
      1,
      `${caseAlias} Message One total isn't same as expected`
    )
    assert.strictEqual(
      msgOneIndex,
      0,
      `${caseAlias} Message One index isn't same as expected`
    )

    // eslint-disable-next-line no-unused-vars
    const [msgTwoDate, id2, msgTwoTotal, msgTwoIndex] = parseId(
      socket.messages[1].subarray(0, ID_SIZE)
    )

    assert.strictEqual(
      msgTwoTotal,
      1,
      `${caseAlias} Message Two total isn't same as expected`
    )
    assert.strictEqual(
      msgTwoIndex,
      1,
      `${caseAlias} Message Two index isn't same as expected`
    )

    assert.deepStrictEqual(
      id1,
      id2,
      `${caseAlias} All chunks should share same ID`
    )
    assert.deepStrictEqual(
      msgOneDate,
      msgTwoDate,
      `${caseAlias} All chunks should share same Date`
    )

    assert.deepStrictEqual(
      Buffer.concat([
        socket.messages[0].subarray(ID_SIZE),
        socket.messages[1].subarray(ID_SIZE)
      ]),
      payload,
      `${caseAlias} received united message should be the same as sent one`
    )

    await socket.stop()

    console.log(`${caseAlias} passed`)
  }

  const errors = tryCountErrorHook()

  await errors.try(() => testClientSmall(DEFAULT_PORT, false))
  await errors.try(() => testClientSmall(DEFAULT_PORT, true))
  await errors.try(testClientLarge)

  if (errors.count === 0) {
    console.log('[client.js] All test for passed\n')
  } else {
    console.log(`[client.js] Has ${errors.count} errors`)
  }

  return errors.count
}

export default clientTest


cjs/index.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// index.js
var e_udp_socket_exports = {};
__export(e_udp_socket_exports, {
  DEFAULT_PORT: () => DEFAULT_PORT,
  UDPClient: () => client_default,
  UDPSocket: () => socket_default,
  WARNING_DECRYPTION_FAIL: () => WARNING_DECRYPTION_FAIL,
  WARNING_MISSING_MESSAGE: () => WARNING_MISSING_MESSAGE,
  _constants: () => constants_exports,
  _identifier: () => identifier_exports
});
module.exports = __toCommonJS(e_udp_socket_exports);

// src/socket.js
var import_socket_udp = require("socket-udp");

// src/collector.js
var import_node_events = __toESM(require("node:events"), 1);

// src/constants.js
var constants_exports = {};
__export(constants_exports, {
  DEFAULT_DECRYPT_FUNCTION: () => DEFAULT_DECRYPT_FUNCTION,
  DEFAULT_ENCRYPT_FUNCTION: () => DEFAULT_ENCRYPT_FUNCTION,
  DEFAULT_PORT: () => DEFAULT_PORT,
  IV_SIZE: () => IV_SIZE,
  WARNING_DECRYPTION_FAIL: () => WARNING_DECRYPTION_FAIL,
  WARNING_MISSING_MESSAGE: () => WARNING_MISSING_MESSAGE
});
var import_node_crypto = __toESM(require("node:crypto"), 1);
var DEFAULT_PORT = 44302;
var IV_SIZE = 16;
var DEFAULT_ENCRYPTION = "aes-256-ctr";
var DEFAULT_ENCRYPT_FUNCTION = (secret, payload) => {
  const iv = import_node_crypto.default.randomBytes(IV_SIZE).subarray(0, IV_SIZE);
  const cipher = import_node_crypto.default.createCipheriv(DEFAULT_ENCRYPTION, secret, iv);
  const beginChunk = cipher.update(payload);
  const finalChunk = cipher.final();
  return Buffer.concat(
    [iv, beginChunk, finalChunk],
    IV_SIZE + beginChunk.length + finalChunk.length
  );
};
var DEFAULT_DECRYPT_FUNCTION = (secret, buffer) => {
  const iv = buffer.subarray(0, IV_SIZE);
  const payload = buffer.subarray(IV_SIZE);
  const decipher = import_node_crypto.default.createDecipheriv(DEFAULT_ENCRYPTION, secret, iv);
  const beginChunk = decipher.update(payload);
  const finalChunk = decipher.final();
  return Buffer.concat(
    [beginChunk, finalChunk],
    beginChunk.length + finalChunk.length
  );
};
var WARNING_MISSING_MESSAGE = Symbol("missing_message");
var WARNING_DECRYPTION_FAIL = Symbol("decryption_fail");

// src/collector.js
var ArrayIndexed = class extends Array {
  size = 0;
  contentLength = 0;
  set(index, value) {
    if (value === void 0)
      return;
    if (this[index] === void 0) {
      this[index] = value;
      this.contentLength += value.length;
      ++this.size;
    }
    return this.size;
  }
  clear() {
    this.length = 0;
  }
};
var Collector = class extends import_node_events.default {
  /** @type {CollectorMap} data */
  #collector = /* @__PURE__ */ new Map();
  /** @type {number} */
  #gcIntervalId;
  /** @type {number} */
  #gcIntervalTime;
  /** @type {number} */
  #gcExpirationTime;
  #gcFunction = () => {
    const dateNow = Date.now();
    for (const [id, payload] of this.#collector) {
      if (payload[1] + this.#gcExpirationTime < dateNow) {
        this.#collector.delete(id);
        this.emit("warning", {
          type: WARNING_MISSING_MESSAGE,
          id,
          date: payload[2]
        });
      }
    }
  };
  constructor({
    gcIntervalTime = 5e3,
    gcExpirationTime = 1e4,
    ...eventEmitterOptions
  }) {
    super(eventEmitterOptions);
    this.#gcIntervalTime = gcIntervalTime;
    this.#gcExpirationTime = gcExpirationTime;
  }
  start() {
    this.#collector.clear();
    this.#gcIntervalId = setInterval(this.#gcFunction, this.#gcIntervalTime);
  }
  stop() {
    clearInterval(this.#gcIntervalId);
    this.#collector.clear();
  }
  /**
   * @param {string} id
   * @param {CollectorElem} payload
   * @return {Map<string, CollectorElem>}
   */
  set(id, payload) {
    return this.#collector.set(id, payload);
  }
  /**
   * @param {string} id
   * @return {CollectorElem}
   */
  get(id) {
    return this.#collector.get(id);
  }
  /**
   * @param {string} id
   * @return {boolean}
   */
  delete(id) {
    return this.#collector.delete(id);
  }
};
var collector_default = Collector;

// src/identifier.js
var identifier_exports = {};
__export(identifier_exports, {
  COUNTER_INDEX_SIZE: () => COUNTER_INDEX_SIZE,
  COUNTER_TOTAL_SIZE: () => COUNTER_TOTAL_SIZE,
  DATE_SIZE: () => DATE_SIZE,
  ID_SIZE: () => ID_SIZE,
  INCREMENTAL_SIZE: () => INCREMENTAL_SIZE,
  RANDOM_SIZE: () => RANDOM_SIZE,
  SEED_N_TOTAL_OFFSET: () => SEED_N_TOTAL_OFFSET,
  SEED_SIZE: () => SEED_SIZE,
  TIME_META_SIZE: () => TIME_META_SIZE,
  generateId: () => generateId,
  parseId: () => parseId,
  setChunkMetaInfo: () => setChunkMetaInfo
});
var import_node_crypto2 = __toESM(require("node:crypto"), 1);
var DATE_SIZE = 6;
var INCREMENTAL_SIZE = 6;
var INCREMENTAL_EDGE = Buffer.alloc(INCREMENTAL_SIZE).fill(255).readUIntBE(0, INCREMENTAL_SIZE);
var TIME_META_SIZE = DATE_SIZE + INCREMENTAL_SIZE;
var RANDOM_SIZE = 6;
var SEED_SIZE = TIME_META_SIZE + RANDOM_SIZE;
var COUNTER_TOTAL_SIZE = 5;
var COUNTER_INDEX_SIZE = 5;
var CHUNK_META_SIZE = COUNTER_TOTAL_SIZE + COUNTER_INDEX_SIZE;
var ID_SIZE = SEED_SIZE + CHUNK_META_SIZE;
var CACHE_SIZE = 2048 * RANDOM_SIZE;
var CACHE_BUFFER = Buffer.alloc(CACHE_SIZE);
var incrementId = 0;
var cacheOffset = 0;
function refreshCache() {
  cacheOffset = 0;
  import_node_crypto2.default.randomFillSync(CACHE_BUFFER, 0, CACHE_SIZE);
}
refreshCache();
function generateId() {
  if (cacheOffset >= CACHE_SIZE)
    refreshCache();
  const id = Buffer.alloc(ID_SIZE);
  id.set(
    [
      CACHE_BUFFER[cacheOffset],
      CACHE_BUFFER[cacheOffset + 1],
      CACHE_BUFFER[cacheOffset + 2],
      CACHE_BUFFER[cacheOffset + 3],
      CACHE_BUFFER[cacheOffset + 4],
      CACHE_BUFFER[cacheOffset + 5]
    ],
    TIME_META_SIZE
  );
  cacheOffset += RANDOM_SIZE;
  incrementId = ++incrementId & INCREMENTAL_EDGE;
  id.writeUIntBE(Date.now(), 0, DATE_SIZE);
  id.writeUIntBE(incrementId, DATE_SIZE, INCREMENTAL_SIZE);
  return id;
}
var SEED_N_TOTAL_OFFSET = SEED_SIZE + COUNTER_TOTAL_SIZE;
function setChunkMetaInfo(id, total, index) {
  id.writeUIntBE(total, SEED_SIZE, COUNTER_TOTAL_SIZE);
  id.writeUIntBE(index, SEED_N_TOTAL_OFFSET, COUNTER_INDEX_SIZE);
}
function parseId(buffer) {
  if (buffer.length !== ID_SIZE)
    throw new Error("id_size_not_valid");
  const date = new Date(buffer.readUintBE(0, DATE_SIZE));
  const id = buffer.subarray(0, SEED_SIZE);
  const total = buffer.readUintBE(SEED_SIZE, COUNTER_TOTAL_SIZE);
  const index = buffer.readUintBE(SEED_N_TOTAL_OFFSET, COUNTER_INDEX_SIZE);
  return [date, id, total, index];
}

// src/socket.js
var UDPSocketPlus = class extends import_socket_udp.UDPSocket {
  /** @type {undefined|function(Buffer): Buffer} */
  #decryptionFunction;
  /** @type {undefined|Buffer} */
  #decryptionSecret;
  /** @type {boolean} */
  #decryptionEnabled = false;
  /** @type {boolean} */
  #fragmentation = true;
  /** @type {CollectorInstance} */
  #collector;
  /** @type {(WarningMessage) => boolean} */
  #collectorWarningHandler = (message) => this.emit("warning", message);
  /**
   * @param {UDPSocketOptions} [options]
   */
  constructor({
    decryption,
    fragmentation = true,
    gcIntervalTime = 5e3,
    gcExpirationTime = 1e4,
    ...udpSocketOptions
  } = {}) {
    super(udpSocketOptions);
    this.#collector = new collector_default({ gcIntervalTime, gcExpirationTime });
    this.#fragmentation = fragmentation;
    if (decryption) {
      if (typeof decryption === "string") {
        this.#decryptionSecret = Buffer.from(decryption, "hex");
        this.#decryptionFunction = (data) => DEFAULT_DECRYPT_FUNCTION(this.#decryptionSecret, data);
      } else if (decryption instanceof Function) {
        this.#decryptionFunction = decryption;
      }
      if (this.#decryptionFunction instanceof Function) {
        this.#decryptionEnabled = true;
      }
    }
  }
  _construct(callback) {
    var _a, _b, _c, _d;
    (_b = (_a = this.#collector).start) == null ? void 0 : _b.call(_a);
    (_d = (_c = this.#collector).on) == null ? void 0 : _d.call(_c, "warning", this.#collectorWarningHandler);
    super._construct(callback);
  }
  _destroy(error, callback) {
    var _a, _b, _c, _d;
    super._destroy(error, callback);
    (_b = (_a = this.#collector).off) == null ? void 0 : _b.call(_a, "warning", this.#collectorWarningHandler);
    (_d = (_c = this.#collector).stop) == null ? void 0 : _d.call(_c);
  }
  /**
   * @param {Buffer} body
   * @param {MessageHead} head
   */
  handleMessage(body, head) {
    if (this.#decryptionEnabled) {
      try {
        body = this.#decryptionFunction(body);
        head.originSize = head.size;
      } catch (e) {
        this.emit("warning", { type: WARNING_DECRYPTION_FAIL });
      }
    }
    if (this.#fragmentation !== true) {
      head.size = body.byteLength;
      return super.handleMessage(body, head);
    }
    const [date, idBuffer, total, index] = parseId(body.subarray(0, ID_SIZE));
    const id = idBuffer.toString("hex");
    let data = this.#collector.get(id);
    if (!data) {
      data = [new ArrayIndexed(total + 1), Date.now(), date, id, 0];
      this.#collector.set(id, data);
    }
    data[0].set(index, body.subarray(ID_SIZE));
    if (this.#decryptionEnabled) {
      data[4] += head.originSize;
    } else {
      data[4] += body.byteLength;
    }
    if (data[0].size === total + 1) {
      this.#collector.delete(id);
      const compiledBody = this.#compileMessage(data[0]);
      head.originSize = data[4];
      head.size = compiledBody.length;
      head.id = id;
      head.sentDate = date;
      return super.handleMessage(compiledBody, head);
    }
    return this.allowPush;
  }
  /**
   * @param {ArrayIndexed<Buffer>} body
   * @return {Buffer}
   */
  #compileMessage(body) {
    let bodyBuffered;
    if (body.length > 1) {
      bodyBuffered = Buffer.alloc(body.contentLength);
      for (let offset = 0, i = 0; i < body.length; ++i) {
        bodyBuffered.set(body[i], offset);
        offset += body[i].length;
      }
    } else {
      bodyBuffered = body[0];
    }
    return bodyBuffered;
  }
};
var socket_default = UDPSocketPlus;

// src/client.js
var import_socket_udp2 = require("socket-udp");
var UDPClientPlus = class extends import_socket_udp2.UDPClient {
  /** @type {boolean} */
  #fragmentation = true;
  /** @type {number} */
  #packetSize;
  /** @type {boolean} */
  #encryptionEnabled = false;
  /** @type {(Buffer) => Buffer} */
  #encryptionFunction;
  /** @type {Buffer} */
  #encryptionSecret;
  /** @type {[Buffer, BufferEncoding][]} */
  #interruptedData = [];
  #sendInterrupted = () => {
    let i;
    for (i = 0; i < this.#interruptedData.length; ++i) {
      super.write(
        this.#interruptedData[i][0],
        this.#interruptedData[i][1]
      );
      if (this.allowWrite === false)
        break;
    }
    this.#interruptedData.splice(0, i + 1);
    if (this.#interruptedData.length > 0) {
      this.once("_drain:internal", this.#sendInterrupted);
    } else if (this.allowWrite) {
      super.emit("drain");
    }
  };
  emit() {
    if (arguments[0] === "drain" && this.#interruptedData.length > 0) {
      return super.emit("_drain:internal");
    }
    return super.emit.apply(this, arguments);
  }
  /**
   * @param {UDPClientOptions} [options]
   */
  constructor({
    encryption,
    fragmentation = true,
    packetSize = 1280,
    ...udpClientOptions
  } = {}) {
    super(udpClientOptions);
    this.#packetSize = packetSize - ID_SIZE;
    this.#fragmentation = fragmentation;
    if (encryption) {
      if (typeof encryption === "string") {
        this.#packetSize = packetSize - IV_SIZE;
        this.#encryptionSecret = Buffer.from(encryption, "hex");
        this.#encryptionFunction = (data) => DEFAULT_ENCRYPT_FUNCTION(this.#encryptionSecret, data);
      } else if (encryption instanceof Function) {
        this.#encryptionFunction = encryption;
      }
      if (this.#encryptionFunction instanceof Function) {
        this.#encryptionEnabled = true;
      }
    }
  }
  write(buffer, encoding, cb) {
    if (this.#fragmentation) {
      this.#sendInFragments(buffer, generateId(), encoding, cb);
    } else {
      if (this.#encryptionEnabled) {
        super.write(
          this.#encryptionFunction(buffer),
          encoding,
          cb
        );
      } else {
        super.write(buffer, encoding, cb);
      }
    }
    if (this.allowWrite === false && this.#interruptedData.length > 0) {
      this.once("_drain:internal", this.#sendInterrupted);
    }
    return this.allowWrite;
  }
  /**
   * @param {Buffer} payload
   * @param {Buffer} id
   * @param {BufferEncoding} encoding
   * @param {function} callback
   * @returns {boolean}
   */
  #sendInFragments(payload, id, encoding, callback) {
    const total = Math.ceil(payload.length / this.#packetSize) - 1;
    const promises = [];
    const callbackExists = callback instanceof Function;
    const createFragmentCallbackPromise = (fragment) => {
      return new Promise((resolve, reject) => {
        super.write(fragment, encoding, (err) => {
          if (err)
            reject(err);
          resolve();
        });
      });
    };
    for (let i = 0; i < payload.length; i += this.#packetSize) {
      let fragment = this.#markFragment(
        id,
        total,
        i / this.#packetSize,
        payload.subarray(i, i + this.#packetSize)
      );
      if (this.#encryptionEnabled) {
        fragment = this.#encryptionFunction(fragment);
      }
      if (this.allowWrite) {
        if (callbackExists) {
          promises.push(createFragmentCallbackPromise(fragment));
        } else {
          super.write(fragment, encoding);
        }
      } else {
        this.#interruptedData.push([fragment, encoding]);
      }
    }
    if (callbackExists) {
      Promise.all(promises).then(() => callback()).catch(callback);
    }
    return this.allowWrite;
  }
  /**
   * @param {Buffer} id
   * @param {number} total
   * @param {number} index
   * @param {Buffer} chunk
   * @returns {Buffer}
   */
  #markFragment(id, total, index, chunk) {
    const resultChunk = Buffer.alloc(chunk.length + ID_SIZE);
    resultChunk.set(id, 0);
    setChunkMetaInfo(resultChunk, total, index);
    resultChunk.set(chunk, ID_SIZE);
    return resultChunk;
  }
};
var client_default = UDPClientPlus;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  DEFAULT_PORT,
  UDPClient,
  UDPSocket,
  WARNING_DECRYPTION_FAIL,
  WARNING_MISSING_MESSAGE,
  _constants,
  _identifier
});
